lib/kube/proxy/resource_filters.go

```diff
@@ -0,0 +1,674 @@
+			negotiator:       negotiator,
+			allowedResources: allowedResources,
+			deniedResources:  deniedResources,
+			log:              log,
+			kind:             kind,
+		}, nil
+	}
+}
+
+// wildcardFilter is a filter that matches all pods.
+var wildcardFilter = types.KubernetesResource{
+	Kind:      types.KindKubePod,
+	Namespace: types.Wildcard,
+	Name:      types.Wildcard,
+}
+
+// containsWildcard returns true if the list of resources contains a wildcard filter.
+func containsWildcard(resources []types.KubernetesResource) bool {
+	for _, r := range resources {
+		if r.Kind == wildcardFilter.Kind &&
+			r.Name == wildcardFilter.Name &&
+			r.Namespace == wildcardFilter.Namespace {
+			return true
+		}
+	}
+	return false
+}
+
+// resourceFilterer is a resource filterer instance.
+type resourceFilterer struct {
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+	// contentType is the response "Content-Type" header.
+	contentType string
+	// responseCode is the response status code.
+	responseCode int
+	// negotiator is an instance of a client negotiator.
+	negotiator runtime.ClientNegotiator
+	// allowedResources is the list of kubernetes resources the user has access to.
+	allowedResources []types.KubernetesResource
+	// deniedResources is the list of kubernetes resources the user must not access.
+	deniedResources []types.KubernetesResource
+	// log is the logger.
+	log logrus.FieldLogger
+	// kind is the type of the resource.
+	kind string
+}
+
+// FilterBuffer receives a byte array, decodes the response into the appropriate
+// type and filters the resources based on allowed and denied rules configured.
+// After filtering them, it serializes the response and dumps it into output buffer.
+// If any error occurs, the call returns an error.
+func (d *resourceFilterer) FilterBuffer(buf []byte, output io.Writer) error {
+	// decode the response into the appropriate Kubernetes API type.
+	obj, bf, err := d.decode(buf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// if bf is not empty, it means that response does not contain any valid response
+	// and it should be safe to write it back into the buffer.
+	if len(bf) > 0 {
+		_, err = output.Write(buf)
+		return trace.Wrap(err)
+	}
+
+	switch allowed, isList, err := d.FilterObj(obj); {
+	case err != nil:
+		return trace.Wrap(err)
+	case isList:
+	case !allowed:
+		// if the object is not a list and it's not allowed, then we should
+		// return an error.
+		return trace.AccessDenied("access denied")
+	}
+	// encode the filterer response back to the user.
+	return d.encode(obj, output)
+}
+
+func arrayToPointerArray[T any](arr []T) []*T {
+	out := make([]*T, len(arr))
+	for i := range arr {
+		out[i] = &arr[i]
+	}
+	return out
+}
+
+func pointerArrayToArray[T any](arr []*T) []T {
+	out := make([]T, len(arr))
+	for i := range arr {
+		if arr[i] == nil {
+			continue
+		}
+		out[i] = *arr[i]
+	}
+	return out
+}
+
+// FilterObj receives a runtime.Object type and filters the resources on it
```

Comment:
Can you document what the second `bool` return does please?

Reply:
Addressed in [b63761b](https://github.com/gravitational/teleport/pull/26244/commits/b63761b7ac33d634e7ad99eca05eb170e0a3dbbe)